src/main/java/org/sagebionetworks/bridge/services/PermissionService.java

```diff
@@ -0,0 +1,169 @@
+        permission.setEntityId(existingPermission.getEntityId());
+        
+        Permission updatedPermission = permissionDao.updatePermission(appId, permission);
+    
+        return getPermissionDetail(appId, updatedPermission);
+    }
+    
+    public List<PermissionDetail> getPermissionsForUser(String appId, String userId) {
+        checkArgument(isNotBlank(appId));
+        checkArgument(isNotBlank(userId));
+    
+        List<Permission> permissions = permissionDao.getPermissionsForUser(appId, userId);
+    
+        List<PermissionDetail> permissionDetails = new ArrayList<>();
+        for (Permission permission : permissions) {
+            permissionDetails.add(getPermissionDetail(appId, permission));
+        }
+    
+        return permissionDetails;
+    }
+    
+    public List<PermissionDetail> getPermissionsForEntity(String appId, String entityType, String entityId) {
+        checkArgument(isNotBlank(appId));
+        checkNotNull(entityType);
+        checkArgument(isNotBlank(entityId));
+        
+        List<Permission> permissions = permissionDao.getPermissionsForEntity(appId, EntityType.valueOf(entityType), entityId);
+        
+        List<PermissionDetail> permissionDetails = new ArrayList<>();
+        for (Permission permission : permissions) {
+            permissionDetails.add(getPermissionDetail(appId, permission));
+        }
+        
+        return permissionDetails;
+    }
+    
+    public void deletePermission(String appId, String guid) {
+        checkArgument(isNotBlank(appId));
+        checkArgument(isNotBlank(guid));
+        
+        permissionDao.deletePermission(appId, guid);
+    }
+    
+    protected PermissionDetail getPermissionDetail(String appId, Permission permission) {
+        AccountId accountId = AccountId.forId(appId, permission.getUserId());
+        Account userAccount = accountService.getAccount(accountId)
+                .orElseThrow(() -> new EntityNotFoundException(Account.class));
+        AccountRef userAccountRef = new AccountRef(userAccount);
+        
+        String entityName = "";
+        if (orgTypes.contains(permission.getEntityType())) {
+            Organization organization = organizationService.getOrganization(appId, permission.getEntityId());
+            entityName = organization.getName();
+        } else if (studyTypes.contains(permission.getEntityType())) {
+            Study study = studyService.getStudy(appId, permission.getEntityId(), false);
+            entityName = study.getName();
+        } else if (assessmentTypes.contains(permission.getEntityType())) {
+            Assessment assessment = assessmentService.getAssessmentByGuid(appId, null, permission.getEntityId());
+            entityName = assessment.getTitle();
+        }
+    
+        EntityRef entityRef = new EntityRef(permission.getEntityType(), permission.getEntityId(), entityName);
+        
+        return new PermissionDetail(permission, entityRef, userAccountRef);
```

Comment:
See comments about not duplicating some information about the permission alongside the entity and account references.

Reply:
Removed the userId, entityType, and entityId fields that are duplicated by the reference types.